[Create Test Module] Finalize and Refactor TC_IDM_2_2 Python3 Test Module

j-ororke opened the pull request:
Description:
This PR finalizes the Python3 test module TC_IDM_2_2.py, building on prior implementation work by Austin. His PR served as the basis for this expanded and corrected version. The test logic has been brought up to spec to ensure proper handling of DataVersion filtering, ReadAttribute operations, and expected error handling.

#### Summary
**Refactor & Alignment with Proven Patterns:**
- Converted test logic to use self.step() for improved step tracking and readability.

**Updated test steps 13–14:**
- Replaced Read with ReadAttribute for consistency and correct response parsing.
- Extracted data version using Clusters.Attribute.DataVersion instead of hardcoded values.
- Aligned assertions to expect {} when the DataVersion matches, per SDK behavior.
- Simplified and unified response handling logic.

**Expanded and Finalized Steps (14–22):**
- Got test steps 14–22 working as expected using corrected and modernized logic:
- WriteAttribute fix: Migrated from deprecated 4-parameter format to correct 2-parameter form (endpoint, attribute(value=...)).
- Error handling for non-global attributes: Now correctly uses attribute_id and expects ChipStackError with 0x580.
- Attribute/event reads: Switched from subscription to Read(nodeid=..., attributes=[()], events=[()]) to better match expected behavior and parse responses accurately.

#### References
**Austin’s original PR (reference implementation):**
- https://github.com/project-chip/connectedhomeip/pull/34003

**Test plan change PR (removes test steps 9–18):**
- https://github.com/CHIP-Specifications/chip-test-plans/pull/5244

**Tracking Issue Task:**
- https://github.com/project-chip/matter-test-scripts/issues/266

#### Acknowledgments
- Thanks to Austin for his initial implementation, which served as the foundation for this work.

#### Testing
- Created TC_IDM_2_2 python3 test module

j-ororke committed:
[Adding] Add TC_IDM_2_2 test module - Update of Austin's module

j-ororke committed:
[TC_IDM_2_2] Refactor data version filter steps to use ReadAttribute and updated test steps to work with TH:

- updated test step logic to using self.step()
- Switched data version filter operations (steps 13–14) to use `ReadAttribute` instead of `Read` for consistency and correct response handling.
- Updated data version extraction to use `Clusters.Attribute.DataVersion` instead of hardcoded attribute IDs.
- Adjusted test assertions to expect an empty dict `{}` when the DataVersion matches, aligning with Matter SDK and Austin's implementation.
- Simplified and unified response handling for filtered reads.
- Ensured test steps 13 and 14 now function as expected per the test plan and Matter spec.

This brings the test logic in line with a proven working pattern and resolves issues with response type mismatches and assertion failures.

j-ororke committed:
Updating IDM_2_2 python3 test module:
- Got test step 14-22 working as expected using the following logic:
- Fixed WriteAttribute API usage: Updated from deprecated 4-parameter format to correct 2-parameter format (endpoint, attribute(value=...)) for data version filter operations and ACL modifications
- Corrected non-global attribute testing: Use attribute_id and expect ChipStackError with code 0x580 for proper error handling
- Updated events/attributes reading: Changed from subscription approach to Read(nodeid=..., attributes=[()], events=[()]) for proper response object handling

j-ororke committed:
Updating Test Step 18 logic and removed some unneeded debug commits
- Updated test step 18 to using the ListLongOctetString attribute from the UnitTesting cluster as mentioned by Gemini bot, however we do not get a SuppressResponse field in the respond back currently
- Removed some previous debug comments in the code that are no longer needed

j-ororke committed:
Minor update to logic for test step 18

j-ororke committed:
Updating _test_read_operation() in TC_IDM_2_2 python3 test module:
- _test_read_operation method and all its operation types have now been refactored into individual async methods
- A dictionary maps each operation type to its handler.
- main _test_read_operation method is now clean and simply dispatches to the correct handler.

j-ororke committed:
Updating TC_IDM_2_2 python3 test module:
- Updated test step 11 to resolve issue with unsupported attribute with for-loop having false positive when not finding an unsupported attribute
- Updated test step 18 to using new logic copied from TCP_Tests.py for sending large data over TCP during testing
- Removed Exception error handling and now using more specific defined exceptions

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,960 @@
+
+        # Step 18: Test chunked data message
+        self.step(18)
+        read_request18 = await self._test_read_operation(operation_type='chunked_data')
```
Currently unable to establish verification logic unless we implement C++ binding in order to access the SuppressResponse field from chunked messages as they are not accessible.

j-ororke committed:
autopep8 and isort fixes

j-ororke committed:
Resolving linting errors

j-ororke committed:
Resolving further style issues

j-ororke committed:
Updating TC_IDM_2_2 python3 test module:
- Updated to using setup_class_method instead of using prior debugging setup class method()

j-ororke committed:
Resolving further linting errors

j-ororke committed:
autopep8 fix

j-ororke committed:
resolving more linting errors; Odd ChipStackError requires # chipstack-ok now to pass linter

j-ororke committed:
resolving more autopep8 errors

j-ororke committed:
sorting endpoint lists for test steps 1 and 2

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

j-ororke committed:
Update TC_IDM_2_2.py

Removing some unnecessary comments from the code

j-ororke committed:
applying isort and autopep8 to latest changes

cecille reviewed (commented):
I haven't quite got into the test part, but can you change the function apis to not use kwargs and pass the required parameters explicitly.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+            'multiple_data_version_filters': self._read_multiple_data_version_filters,
+        }
+
+    def get_device_clusters(self) -> set:
```
Ah, thank you, you are correct. The manual discovery of endpoints, clusters, and attributes had been removed as the setup class helper here {await self.setup_class_helper(allow_pase=False)} does that for us, without having to run these individual methods like was done previously it appears.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+        self.large_operation_timeout_ms = 60000  # 60 second timeout for large operations (like reading all attributes)
+        self.setup_timeout_ms = 60000  # 60 second timeout for setup phase
+        # Operation type to handler mapping
+        self._operation_map = {
```
Thank you for the suggestion: That’s a great point. I’ve updated the code to pass the function references directly through the dictionary, so we’re no longer using the string map approach.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+
+    # === Attribute Reading and Verification ===
+
+    async def read_attribute(self, attribute_path: list, timeout_ms: int = None) -> dict:
```
Thank you for pointing this out! 
You’re right: the timeout_ms param is currently set, but not actually used to enforce a timeout on the read operation. This was likely intended for future-proofing, but as it stands, it doesn’t affect the read. I’ll either remove it as I don't currently see a good reason to keep it.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+                        attribute=attribute_path[0].AttributeId,
+                        error=expected_error)
+                else:
+                    # Handle tuple paths
```
Hi Cecille, great catch, thank you! 
This unpacking logic was there originally to support test steps involving invalid endpoints, clusters, and attributes. I’ve now updated the code to unpack these directly, and it works correctly for all test step scenarios.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+            AssertionError if verification fails
+        """
+        for endpoint in read_request.tlvAttributes:
+            # Verify parts list if present
```
Good point, thank you! I’ll update the code to assert that the parts list is always present.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+                              read_request.tlvAttributes[endpoint][Clusters.Descriptor.id],
+                              "AttributeList not in output")
+
+    async def _read_single_attribute(self, **kwargs):
```
Hi Cecille,
Thank you and great catch, you’re absolutely right: using **kwargs and string maps here is unnecessary and can be error-prone. I’ll refactor these methods to use explicit parameters instead, which will make the code safer and clearer.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+            AssertionError if verification fails
+        """
+        if hasattr(cluster_obj, 'cluster_id'):
+            cluster = ClusterObjects.ALL_CLUSTERS[cluster_obj.cluster_id]
```
Hi Cecille, 
So good question, the test steps that use this function all have the Descriptor cluster. 
So going by the YAGNI (You Aren't Gonna Need It) principle. I have removed unnecessary complexity because we know that this method will only use with the Descriptor cluster. Making the code here now much cleaner and more maintainable.

j-ororke committed:
Updated TC_IDM_2_2 python3 test module to resolve comments from Cecille:

- Removed timeout_ms parameter from read_attribute()
- Removed manual discovery functions for endpoints, clusters, and attributes as already gathered with setup_class_helper(allow_pase=False) at the start of the test
- Removed string map established for operation_type dictionary and now passing directly to the functions established in the dictionary, no more string maps.
- Changed to using direct unpacking of endpoint, cluster, and attributes in attribute read and expect error scenarios.
- Parts list is always present, but added assertion instead of a if block for quick and simple verification.
- Updated _verify_single_endpoint_path() to include type
- Removed kwargs in private functions to now pass in the variables directly. String maps are dangerous and should be avoided.
- Updated to remove cluster_obj var and parameter from async function _verify_single_endpoint_path() as we know this test currently only uses the Descriptor cluster with this function, so we can just set the cluster in the function itself and remove any complexities using the YAGNI principle

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,891 @@
+                returned_attrs = sorted([x for x in read_request.tlvAttributes[endpoint][cluster].keys()])
+                attr_list = sorted([x for x in read_request.tlvAttributes[endpoint][cluster][
+                    ClusterObjects.ALL_CLUSTERS[cluster].Attributes.AttributeList.attribute_id]
+                    if x != Clusters.UnitTesting.Attributes.WriteOnlyInt8u.attribute_id])
```
Understood, will discuss this with Amine when possible.

j-ororke committed:
Update TC_IDM_2_2.py to resolve comments from Cecille

- Removed the timeout_ms from the read_attribute async function

j-ororke committed:
Resolving linting issues

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

j-ororke committed:
Updating chip -> matter for imports

j-ororke committed:
Restylized by isort

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+        self.large_operation_timeout_ms = 60000  # 60 second timeout for large operations (like reading all attributes)
+        self.setup_timeout_ms = 60000  # 60 second timeout for setup phase
+        # Operation type to handler mapping (use function references directly)
+        self._operation_map = {
```
Ah, you’re right — it’s effectively mapping functions to themselves right now, which doesn’t add value. I’ll simplify this by removing the _operation_map and just use the functions directly.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+        super().__init__(*args, **kwargs)
+        self.endpoint = 0
+        self.timeout_ms = 10000  # 10 second timeout for read operations
+        self.large_operation_timeout_ms = 60000  # 60 second timeout for large operations (like reading all attributes)
```
Sorry, this was leftover when trying to get all endpoints/clusters/attributes from the DUT. 
I had not realized we already had the BasicComposition classes self.setup_class_helper() function for this in basic_composition test support module.

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.endpoint = 0
+        self.timeout_ms = 10000  # 10 second timeout for read operations
```
Thank you, it appears the need for this has also been removed. 
So I apologize I must have missed cleaning this up, will remove it here now

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+        """Read and verify attributes from the device.
+
+        Args:
+            attribute_path: List of attribute paths to read (should be AttributePath or direct values)
```
Updated this docstring line to this here now as it provided a clearer understanding and removes the prior ambiguousness:
attribute_path: List of attribute paths to read (should be AttributePath objects or tuples like (endpoint, cluster, attribute))

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+        else:
+            await self._verify_single_endpoint_path(read_request, endpoint, attribute)
+
+    async def _verify_single_endpoint_path(
```
You're absolutely right! 
I went through and cleaned this up - way too much unnecessary async going on.
Converted to sync (just doing data validation):

- verify_read_response()
- verify_all_endpoints_clusters()
- verify_specific_path()
- _verify_single_endpoint_path()
- verify_cluster_path()
- verify_attribute_path()
- _verify_empty_tuple()
- These were all just doing assertions and data comparisons, no actual device calls.

Kept async (actually talking to the device):

- read_attribute() - calls self.default_controller.Read()
- verify_attribute_read() - orchestrates the reads
- All the _read_* functions anything that makes controller calls
- Test step functions - coordinate multiple device operations

Feels much better now; async only where we actually need it for device communication. 
Thanks for pushing back on this, it's way cleaner!

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+        Raises:
+            AssertionError if verification fails
+        """
+        # This method is only used for Descriptor cluster, so we can use it directly
```
You're absolutely right on both points! 
I've updated the comment and code to be clearer. 
The function is indeed called by verify_specific_path() and the comment was misleading - it's not 'only used for Descriptor cluster' but rather uses Descriptor cluster data to validate any endpoint read. Fixed the comment to accurately describe that it validates endpoint data by checking ServerList regardless of what was originally read.

j-ororke committed:
Refactor TC_IDM_2_2:
- Remove unnecessary vars from test initialization
- Removed prior function mapping that was previously mapping functions to themselves
- Remove unnecessary async from verification functions by converting verification functions to sync (verify_read_response, verify_all_endpoints_clusters, verify_specific_path, _verify_single_endpoint_path, verify_cluster_path, verify_attribute_path, _verify_empty_tuple)
- Keep async only for functions that make device controller calls
- Fix misleading comment in _verify_single_endpoint_path
- Update docstring to clarify AttributePath vs tuple parameters
- Add type checking for AttributePath vs tuple in verify_attribute_read

j-ororke committed:
Restylizing TC_IDM_2_2 python3 test module

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

j-ororke committed:
Update TC_IDM_2_2 python3 test module:
- Simplifies _read_single_attribute by adding cluster parameter and removing unnecessary cluster detection logic
- Updated _read_unsupported_attribute to skip gracefully when no unsupported attributes found (expected with chip-all-clusters-app)
- Added verification logic to _read_repeat_attribute to ensure repeated reads return consistent values
- Removed problematic fallback code that could fail if no thermostat cluster

j-ororke committed:
Restylizing TC_IDM_2_2 python3 test module

j-ororke committed:
Updating test step 5 ret and exp attrs due to CI failure on Clusters.UnitTesting.Attributes.WriteOnlyInt8u.attribute_id

j-ororke committed:
Restylizing TC_IDM_2_2 python3 test module

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+
+    # === Attribute Reading and Verification ===
+
+    async def read_attribute(self, attribute_path: list) -> dict:
```
OK, that makes sense, but I'm still not totally sure what the purpose of this function is. It abstracts out the dut_node_id and the default controller, but this only called once and it's really just re-logging the attribute path, which is logged before this call. I think you can get rid of this function, call Read directly in the one place this function is used, and not lose anything. It'll make the code easier to follow.

j-ororke replied:
No problem, I have removed the prior read_attribute function and now have the ReadAttribute function called directly inside the verify_attribute_read function.

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,733 @@
+                    endpoint = attribute_path[0][0] if len(attribute_path[0]) > 0 else None
+                    cluster = attribute_path[0][1] if len(attribute_path[0]) > 1 else None
+                    attribute = attribute_path[0][2] if len(attribute_path[0]) > 2 else None
+                result = await self.read_single_attribute_expect_error(
```
You'll get a KeyError here if you pass in anything other than a single attribute. I don't think this code is working as expected.

read_single_attribute_expect_error uses the endpoints as keys to read the returned attribute paths. This will throw a key error if the key is None. That doesn't mean the ReadAttribute call failed with the expected error.

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,733 @@
+                    attribute = attribute_path[0].AttributeId
+                else:
+                    # Handle tuple case
+                    endpoint = attribute_path[0][0] if len(attribute_path[0]) > 0 else None
```
Instead of parsing these from the tuple, why not just pass in the AttributePath? You can use it directly.

j-ororke replied:
So this has actually been removed here now as previously the unsupported endpoint and cluster were run through this function. However, after you mentioned above that the logic was not correct. I went back to the drawing board and updated the _verify_unsupported_endpoint and _verify_unsupported_cluster functions. 
I found that indeed the prior method had some flaws and were providing back a KeyError you mentioned instead of the InteractionModel error that we were wanting, so I have updated it now to using a proper method for both of those functions here now to filter through all the endpoints and clusters, then either add onto the endpoints to make a new one that doesn't exist, or filter through the clusters until I find one that isnt on an endpoint that the DUT has supported, so this block of code no longer exists.

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,733 @@
+        self.verify_read_response(read_request, attribute_path)
+        return read_request
+
+    def verify_read_response(self, read_request: dict, attribute_path: list):
```
```suggestion
    def verify_read_response(self, read_response: dict, attribute_path: list):

```

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,733 @@
+
+        if isinstance(attribute_path[0], tuple):
+            self.verify_specific_path(read_request, attribute_path[0])
+        elif isinstance(attribute_path[0], type):
```
What is "type" here. 

I think your life might be easier if you just accept AttributePath and drop tuple support.

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,733 @@
+                "PartsList attribute should always be present")
+            parts_list = read_request.tlvAttributes[endpoint][Clusters.Descriptor.id][
+                Clusters.Descriptor.Attributes.PartsList.attribute_id]
+            if Clusters.Descriptor.Attributes.PartsList in self.endpoints[endpoint][Clusters.Descriptor]:
```
is this verifying the parts list against itself? I think the intent of this is to verify that the read response returns values for all the endpoints. The parts list on EP0 is supposed to contain a list of all the endpoints (minus itself). So here, you know you got back data for all the expected endpoints if read_request.tlvAttributes.keys() == parts_list + 0

cecille replied:
That should be outside of the for endpoint... construct

j-ororke replied:
Thank you, this is much better and cleaner now than the circular logic that was in place. The verification now actually checks that the read response returns values for all the expected endpoints based on the parts list gathered from EP0 (adding itself back in in the equation + 0), rather than just verifying the parts list against itself as was done previously.

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,733 @@
+            self.verify_attribute_path(read_request, attribute_path[0])
+
+    def verify_all_endpoints_clusters(self, read_request: dict):
+        """Verify read response for all endpoints and clusters.
```
This is specifically for a full wildcard read, right? because otherwise you won't have a full match across all these checks.

j-ororke replied:
Hi Cecille,
You are correct, this is for a full wildcard read verification only, specifically it is intended for test step 5's "read all attributes from all clusters on all endpoints". Sorry, the information provided in the docstring here was not very clear. I have now updated it to show this better in the event that it needs to be updated or modified in the test module later.

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,733 @@
+            dataVersionFilters=data_version_filter_1)
+        return read_request, filtered_read
+
+    def _verify_empty_tuple(self, attr_path, read_request):
```
```suggestion
    def _verify_full_wildcard(self, attr_path, read_request):

```

j-ororke replied:
Thank you, this has been updated!

cecille commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,768 @@
+                nodeid=self.dut_node_id,
+                allowPASE=False,
+                timeoutMs=1000,
+                payloadCapability=ChipDeviceCtrl.TransportPayloadCapability.LARGE_PAYLOAD
```
I would argue that we should omit this explicit check - if the data came back, we know it was chunked properly because it has to have been. I think we can realistically just remove this step from the test - we do full wildcard reads all the time. In fact, you do it already in this test. That's the max amount of data that's coming back from the device. It's chunked. I can't imagine a case where it wouldn't be with the root node endpoint the size it is, and if it's not...well, nothing's going to be chunked, really. 

Anyway, suggest just deleting this step in the test plan with the explanation that a full wildcard read is known to produce chunked results and the fact that we can parse the data correctly is sufficient proof that this is working. No need to go digging into packets for the sake of it. And there's no need to do it again. It's already done in another step.

j-ororke replied:
Created new test plan PR here to remove current test step 18 from this test.

j-ororke committed:
Update src/python_testing/TC_IDM_2_2.py

Co-authored-by: C Freeman <cecille@google.com>

j-ororke committed:
autopep8 restyles TC_IDM_2_2 test module

j-ororke committed:
Updating TC_IDM_2_2 test module:
- Changed function name from _verify_empty_tuple to _verify_empty_wildcard as suggested by Cecille
- Updated function calls to reflect name change for _verify_empty_wildcard
- Updated read_request arg name to read_response for verify_read_response function call

j-ororke committed:
Updating read_request var name to read_response where needed

j-ororke committed:
Removing test step 18 to comply with test plan PR #[5467](https://github.com/CHIP-Specifications/chip-test-plans/pull/5467)

j-ororke committed:
Resolving linting errors

j-ororke committed:
Refactor TC_IDM_2_2 test module based on Cecille's feedback:
- Remove read_attribute function and replace with direct controller.Read calls
- Fix KeyError in unsupported endpoint/cluster tests by using read_single_attribute_expect_error directly instead of verify_attribute_read
- Add comprehensive logic for unsupported cluster validation that iterates through all standard clusters across all endpoints to find actual unsupported ones
- Implement dynamic unsupported endpoint detection using Austin's approach (find endpoints beyond max+2)
- Simplify verify_attribute_read to only accept AttributePath objects:
  * Remove support for tuple and type formats
  * Standardize all test calls to use AttributePath objects
  * Delete unused verification methods (verify_specific_path, verify_cluster_path)
  * Make API cleaner and more maintainable
- Fix circular logic in verify_all_endpoints_clusters by properly validating that read response contains data for all endpoints listed in EP0's parts list
- Update verify_all_endpoints_clusters docstring to clarify it's only for full wildcard reads (test step 5)
- Remove some intermediary wrapper functions, now calling verify_attribute_read with AttributePath objects directly

j-ororke committed:
isort and autopep8 stylized

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

j-ororke committed:
Updating TC_IDM_2_2 PR test module:
- Removed verify_read_response function entirely - now calls verify_attribute_path directly
- Added comprehensive verification for server list, endpoints, and attribute lists
- Created shared _verify_attributes_match_list function to handle attribute verification across multiple response formats (Read() vs ReadAttribute())
- Updated code to only verify spec-defined clusters using global_attribute_ids.cluster_id_type() to skip clusters that may have write-only attributes
- Fixed response format handling to support both .tlvAttributes (Read) and direct dict (ReadAttribute) formats
- Fixed cluster object vs cluster ID handling in shared verification function
- Added missing endpoint verification to _read_cluster_all_endpoints
- Added missing ServerList verification to _read_endpoint_all_clusters
- Changed function name from _read_non_global_attribute to _read_non_global_attribute_across_all_clusters

j-ororke committed:
stylized

j-ororke committed:
Updating TC_IDM_2_2 test module::
- Fixed global attribute verification logic to check presence in ALL clusters instead of just ANY cluster (lines 182-185)
- Added DataVersion filtering in _read_cluster_all_endpoints to prevent AttributeError when accessing .attribute_id (lines 216-217)
- Fixed AttributeList access logic to properly handle different response formats and prevent KeyError (lines 211-219)
- Updated _read_unsupported_attribute to fail test instead of skipping when no unsupported attributes found, as all clusters should have at least one unsupported attribute (lines 331-332)

j-ororke committed:
Restylized by autopep8

cecille reviewed (approved):
Couple of minor things, but this is good to go once those are fixed up

j-ororke commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,674 @@
+                                          f"AttributeList not found in cluster {cluster_id} on endpoint {ep}")
+
+                        # Verify that returned attributes match the AttributeList
+                        if global_attribute_ids.cluster_id_type(cluster_id) == global_attribute_ids.ClusterIdType.kStandard:
```
Thank you for catching that! 
I had mistakenly applied the same validation logic to all three places where we check attributes, but they actually need different approaches
Now I understand the distinction you were making:

- Wildcard reads (verify_all_endpoints_clusters): These need the conditional check to skip MEI clusters because wildcard reads return write-only attributes that won't be in the AttributeList, causing failures.
- Targeted reads (verify_attribute_path): These don't have write-only attribute issues, so the validation should work for all clusters including MEI ones, as you mentioned.

j-ororke committed:
Fix attribute verification logic and add clarifying comment:
- Removed conditional check for standard clusters in verify_attribute_path since targeted reads don't have write-only attribute issues and should verify all cluster types including MEI clusters
- Added comment explaining that global attribute verification checks for specific global attributes across all clusters

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated

soares-sergio commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,674 @@
+#
+#    Copyright (c) 2024 Project CHIP Authors
```
```suggestion
#    Copyright (c) 2025 Project CHIP Authors
```

j-ororke committed:
Removing sleep and docstring from test step 14

j-ororke committed:
Resolving linting error with unused impoart time

j-ororke committed:
Update src/python_testing/TC_IDM_2_2.py

Changing copyright date from Sergio's suggestion.

Co-authored-by: Sergio Soares <sergiosoares@google.com>

soares-sergio commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,674 @@
+        read_request15, filtered_read15 = await self._read_data_version_filter(
+            endpoint=self.endpoint,
+            cluster=Clusters.BasicInformation,
+            attribute=Clusters.BasicInformation.Attributes.NodeLabel,
```
The test plan spec for step 15 says:
```
TH sends a Read Request Message to the DUT to read all attributes on a cluster with the DataVersionFilter Field not set.
```

From the code, it seems it's reading only `NodeLabel` ?

j-ororke replied:
Hey Sergio!

You're absolutely right - good catch! 👍

I've updated the logic in test step 15 to now read all attributes from the BasicInformation cluster (not just NodeLabel). The test now correctly:

- Reads all attributes on the cluster → captures DataVersion (V1)
- Writes to an attribute → DataVersion changes to V2
- Reads all attributes again with the old DataVersion filter (V1)
- Verifies the DUT returns all attributes with the new DataVersion (V2)

I agree the test step wording is a bit unclear - it says "with the dataversion value received before" which refers to the filter using the old version, but the expected result is that we get back all attributes with the new version (since the version changed). The implementation correctly validates this behavior.

Thanks for catching that!

soares-sergio commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,669 @@
+            )
+            asserts.fail("Expected INVALID_ACTION error but operation succeeded")
+        except ChipStackError as e:  # chipstack-ok
+            asserts.assert_equal(e.err, 0x580,
```
```suggestion
            asserts.assert_equal(e.err, 0x80,
```

Could you double check this code? According to the spec here (https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Interaction-Model.adoc#101-status-code-table) INVALID_ACTION is `0x80` instead of `0x580`

soares-sergio replied:
Oh, I see, thanks! Maybe it would be a good idea to actually list all these error codes in a file with named constants/enum that can be reused? That way, people reading the code, don't have to check that the code is correct, they will see something like `ERROR_CODES.INVALID_ACTION`, etc.

soares-sergio replied:
This can be done as a follow up, though, so don't need to do this now. Just a suggestion that might help all of the test implementations.

soares-sergio commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,669 @@
+        await self._read_non_global_attribute_across_all_clusters(
+            attribute=Clusters.Descriptor.Attributes.ServerList)
+
+        self.step(20)
```
Is this test step trying to read all attributes?

The test plan says `TH sends a Read Request Message to the DUT to read all attributes from all clusters at Endpoint1`

and then `Verify that the DUT sends back data of all attributes only from that one cluster to which it has access.`

I couldn't tell exactly if `_read_limited_access` was doing the right thing. On a quick look, it seems it's not trying to read all attributes... could you double check?

j-ororke replied:
Great catch on Step 20! 
You're right that it wasn't properly testing the limited access behavior upon further inspection.
I've updated _read_limited_access to:

- Create a second controller (TH2) with View access to only one cluster on the endpoint
- Have TH2 read all attributes from all clusters at the endpoint
- Verify only the granted cluster is returned (no errors for others - they're silently filtered)

This now properly demonstrates access control enforcement where a controller with limited permissions only sees the cluster it has access to. Thanks for catching that!

j-ororke committed:
Add comment and spec reference for catching 0x580 (GENERAL_ERROR + INVALID_ACTION)

j-ororke committed:
Fix TC_IDM_2_2 Test Step 15 to read all cluster attributes with data version filter

j-ororke committed:
Update TC_IDM_2_2 Fix Step 16 to send 2 DataVersionIB filters

j-ororke committed:
Update TC_IDM_2_2 test step 16 to remove prior print and exit logic from debugging

j-ororke committed:
Create second controller (TH2) with limited View access to properly test access control enforcement

j-ororke committed:
Updated asserts in test step 20 to more specifically verify we got the correct cluster returned during this test step

j-ororke committed:
Restyled by autopep8

j-ororke committed:
Added check for Clusters.BasicInformation before running test steps 14-17 and test step 20

j-ororke committed:
Update TC_IDM_2_2.py

Updated to include running for a longer time due to possible timeout in highly congested lab networks since this pull all attributes/clusters/endpoints for testing.

j-ororke committed:
Restyled by autopep8

j-ororke committed:
Restyled by autopep8

j-ororke commented:
Latest result from running in TH UI version v2.14-beta2.1+fall2025 locally:
<img width="626" height="776" alt="image" src="https://github.com/user-attachments/assets/222c6ba4-0bfe-4675-adcc-b07e9cb47c43" />

raju-apple commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,777 @@
+                                          f"AttributeList not found in cluster {cluster_id} on endpoint {ep}")
+
+                        # Verify that returned attributes match the AttributeList
+                        returned_attrs = sorted([x for x in read_response.tlvAttributes[ep][cluster_id].keys()])
```
Should we add an assert here if cluster_id doesn't exist in the endpoint ? Looks like we will raise KeyError instead of an assert ?

raju-apple commented on src/python_testing/TC_IDM_2_2.py:
```diff
@@ -0,0 +1,777 @@
+            [(self.endpoint, Clusters.AccessControl.Attributes.Acl(dut_acl_original))])
+        logging.info("Restored original ACL")
+
+        # Removes TH2 controller
```
Guess we will only clean up the controller in case of the happy path , in case of an exception before this we might end up not cleaning up properly.

j-ororke committed:
Updating TC_IDM_2_2 python3 test module to respond to Raju's comments:
- Additional assertion added to ensure cluster_id exists at line 183
- Added try-finally clause to make sure that ACL restoration and TH2 controller occurs even if exceptions happen in _read_limited_access()

j-ororke committed:
Restyled by autopep8

j-ororke committed:
Merge branch 'master' into IDM_2-2_Austins-test-module-updated
